Python/pystate.c

```diff
@@ -1386,28 +1391,62 @@ allocate_chunk(int size_in_bytes, _PyStackChunk* previous)
     return res;
 }
 
-static _PyThreadStateImpl *
-alloc_threadstate(void)
+static void
+reset_threadstate(_PyThreadStateImpl *tstate)
 {
-    return PyMem_RawCalloc(1, sizeof(_PyThreadStateImpl));
+    // Set to _PyThreadState_INIT directly?
+    memcpy(tstate,
+           &initial._main_interpreter._initial_thread,
+           sizeof(*tstate));
+}
+
+#define LOCK_THREADS_FREELIST(interp) \
+    PyMutex_LockFlags(&(interp)->threads.freelist.mutex, _Py_LOCK_DONT_DETACH)
+#define UNLOCK_THREADS_FREELIST(interp) \
+    PyMutex_Unlock(&(interp)->threads.freelist.mutex)
+
+static _PyThreadStateImpl *
+alloc_threadstate(PyInterpreterState *interp)
+{
+    // Try the freelist first.
+    _PyThreadStateImpl *tstate = NULL;
+    LOCK_THREADS_FREELIST(interp);
+    if (interp->threads.freelist.head != NULL) {
+        tstate = interp->threads.freelist.head;
+        interp->threads.freelist.head =
+            (_PyThreadStateImpl *)tstate->base.next;
+    }
+    UNLOCK_THREADS_FREELIST(interp);
+    // Fall back to the allocator.
+    if (tstate == NULL) {
+        tstate = PyMem_RawCalloc(1, sizeof(_PyThreadStateImpl));
+        reset_threadstate(tstate);
+    }
+    return tstate;
 }
 
 static void
 free_threadstate(_PyThreadStateImpl *tstate)
 {
+    PyInterpreterState *interp = tstate->base.interp;
     // The initial thread state of the interpreter is allocated
     // as part of the interpreter state so should not be freed.
-    if (tstate == &tstate->base.interp->_initial_thread) {
-        // Restore to _PyThreadState_INIT.
-        memcpy(tstate,
-               &initial._main_interpreter._initial_thread,
-               sizeof(*tstate));
+    if (tstate == &interp->_initial_thread) {
+        // Add it to the freelist.
+        reset_threadstate(tstate);
```

Comment:
`alloc_threadstate` already resets the thread state, it's probably fine to keep it in the freelist while still initialized (in fact, that will hopefully be helpful in debugging).

Reply:
It resets it in `alloc_threadstate()` only if it is allocated there.  You may be right about keeping the old data being helpful while debugging, but I'm not sure it would make much of a difference.  Regardless, I'll think about it.  Thanks for bringing this up.